jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/MultiAuthenticator.java

```diff
@@ -0,0 +1,457 @@
+        {
+            return _delegate.getUserIdentity();
+        }
+
+        @Override
+        public Principal getUserPrincipal()
+        {
+            return _delegate.getUserPrincipal();
+        }
+
+        @Override
+        public boolean isUserInRole(String role)
+        {
+            return _delegate.isUserInRole(role);
+        }
+
+        @Override
+        public void logout(Request request, Response response)
+        {
+            _delegate.logout(request, response);
+            doLogout(request);
+        }
+    }
+
+    private class DefaultAuthenticator implements Authenticator
+    {
+        @Override
+        public void setConfiguration(Configuration configuration)
+        {
+        }
+
+        @Override
+        public String getAuthenticationType()
+        {
+            return "DEFAULT";
+        }
+
+        @Override
+        public Constraint.Authorization getConstraintAuthentication(String pathInContext, Constraint.Authorization existing, Function<Boolean, Session> getSession)
+        {
+            if (isLoginPage(pathInContext))
+                return Constraint.Authorization.ALLOWED;
+            return existing;
+        }
+
+        @Override
+        public AuthenticationState validateRequest(Request request, Response response, Callback callback)
+        {
+            if (_loginPath != null)
+            {
+                String loginPath = URIUtil.addPaths(request.getContext().getContextPath(), _loginPath);
+                if (_dispatch)
+                {
+                    HttpURI.Mutable newUri = HttpURI.build(request.getHttpURI()).pathQuery(loginPath);
+                    return new AuthenticationState.ServeAs(newUri);
+                }
+                else
+                {
+                    Session session = request.getSession(true);
+                    String redirectUri = session.encodeURI(request, loginPath, true);
+                    Response.sendRedirect(request, response, callback, redirectUri, true);
+                    return AuthenticationState.CHALLENGE;
+                }
+            }
+            return null;
+        }
+    }
+
+    private static MultiAuthState getAuthState(Session session)
+    {
+        if (session == null)
+            return null;
+        return (MultiAuthState)session.getAttribute(AUTH_STATE_ATTR);
+    }
+
+    private static MultiAuthState ensureAuthState(Session session)
+    {
+        if (session == null)
+            throw new IllegalArgumentException();
+
+        MultiAuthState authState = (MultiAuthState)session.getAttribute(AUTH_STATE_ATTR);
+        if (authState == null)
+        {
+            authState = new MultiAuthState();
+            session.setAttribute(AUTH_STATE_ATTR, authState);
+        }
+        return authState;
+    }
+
+    private static boolean isLoggedIn(Session session)
+    {
+        if (session == null)
+            return false;
+
+        synchronized (session)
```

Comment:
This is unnecessary, following the discussion on #10392.

I'd remove all the `synchronized` blocks around the session, as I understand from @janbartel that it is going to be a different instance for every request in any case, so synchronization is useless.